### PR TITLE
feat: DatePicker support value-format

### PR DESCRIPTION
**Pull request by msidolphin**
Please make sure these boxes are checked before submitting your PR, thank you!

reference #1923

* [x] Make sure you follow Element's contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by JeremyWuuuuu** [comment]
Thank you for contributing, seems like you have a member which was not used and the linter complains about it. Please fix it ;)

**Comment by iamkun** [comment]
Nice catch, seems better to add this feature back

File: packages/time-picker/src/common/picker.vue [hunk 1]
```diff
@@ -274,15 +289,18 @@ export default defineComponent({
...
+          result = parser(props.modelValue as Date, props.valueFormat)
         }
       }
 
       if (pickerOptions.value.getRangeAvaliableTime) {
         result = pickerOptions.value.getRangeAvaliableTime(result)
       }
+      if (Array.isArray(result) && result.some(_ => !_)) {
+        result = []
+      }
```
Comment (0song) [inline packages/time-picker/src/common/picker.vue:303]: What's this line used for, please?

File: packages/time-picker/src/common/picker.vue [hunk 2]
```diff
@@ -223,7 +232,13 @@ export default defineComponent({
...
     const emitInput = val => {
       if (!valueEquals(props.modelValue, val)) {
-        ctx.emit('update:modelValue', val)
+        let formatValue
+        if (Array.isArray(val)) {
+          formatValue = val.map(_ => formatter(_, props.valueFormat))
+        } else if(val) {
+          formatValue = formatter(val, props.valueFormat)
+        }
+        ctx.emit('update:modelValue', val ? formatValue : val)
```
Comment (0song) [inline packages/time-picker/src/common/picker.vue:241]: ctx.emit('update:modelValue', val || formatValue)

File: packages/time-picker/src/common/picker.vue (hunk 1 above)
Comment (msidolphin) [reply packages/time-picker/src/common/picker.vue:303]: When the daterange selector used this property and the value is set to an illegal format, the value of parsedValue is [undefined, undefined], and an error will be reported when listening to parsedValue in panel-date-range.vue

当日期范围选择器使用这个属性并且将绑定的value设置为非法格式的时候，parsedValue的值是[undefined, undefined]，在panel-date-range.vue中监听parsedValue的时候会报错

File: packages/time-picker/src/common/picker.vue (hunk 2 above)
Comment (msidolphin) [reply packages/time-picker/src/common/picker.vue:241]: ctx.emit('update:modelValue', formatValue || val)

Comment (0song) [reply packages/time-picker/src/common/picker.vue:241]: what I mean is to simplify it from `val ? formatValue : val` to `val || formatValue`

Comment (msidolphin) [reply packages/time-picker/src/common/picker.vue:241]: when empty case, I did't change the origin way

Comment (msidolphin) [reply packages/time-picker/src/common/picker.vue:241]: for example：click clear button

Comment (msidolphin) [reply packages/time-picker/src/common/picker.vue:241]: actually  `ctx.emit('update:modelValue', val ? formatValue : val)` is the right way

**Comment by msidolphin** [comment]
> Thank you for contributing, seems like you have a member which was not used and the linter complains about it. Please fix it ;)

what issue?

**Comment by 0song** [comment]
> > Thank you for contributing, seems like you have a member which was not used and the linter complains about it. Please fix it ;)
> 
> what issue?

image

**Review by iamkun — APPROVED** [review]
LGTM

**Pull request merged**